Circuit

JohnDoePlusPlus opened the pull request.

Commit:
Circuit

dimadeveatii commented on lib/circuit/CircuitOptions.ts:
```diff
@@ -41,17 +41,15 @@ export type CircuitOptions = {
+export const DEFAULT_POLICY = 'errors';
+export const DEFAULT_ON_ERROR = 'throw';
+export const DEFAULT_SCOPE = 'class';
+export const DEFAULT_ERROR_FILTER = () => true;
```
Maybe we can get rid of these, and just provide defaults directly in the `DEFAULT_OPTIONS`

dimadeveatii commented on lib/utils/isPromiseLike.ts:
```diff
@@ -0,0 +1,3 @@
+export function isPromise(data: any): data is Promise<unknown> {
+  return data && data.then && data.catch;
```
In JS a promise-like is considered any object that has the `then` method. You need to check that there is `then` and that it's a function.

dimadeveatii commented on lib/circuit/factories/circuitStateStorageFactory.ts:
```diff
@@ -0,0 +1,34 @@
+  options: CircuitOptions,
+): CircuitStateStorage {
+
+  const scope = options.scope || DEFAULT_SCOPE;
```
At this point options should already be populated with default.

dimadeveatii commented on lib/circuit/factories/circuitStateFactory.ts:
```diff
@@ -0,0 +1,23 @@
+  const interval = options.interval || DEFAULT_INTERVAL;
+  const errorFilter = options.errorFilter || DEFAULT_ERROR_FILTER;
+
+  const policyType = options.policy || DEFAULT_POLICY;
```
Defaults should be already populated in options at this point.

dimadeveatii commented on lib/circuit/CircuitState/CircuitState.ts:
```diff
@@ -0,0 +1,70 @@
+    const type = isError ? 'error' : 'success';
+
+    if (this.state === 'half-open') {
+      isError ? this.close() : this.open();
```
Ternary operator is recommended with assignments, in this case it has side-effects.

dimadeveatii commented on lib/circuit/CircuitState/CircuitState.ts:
```diff
@@ -0,0 +1,70 @@
+export class CircuitState {
+
+  private state: 'open' | 'close' | 'half-open' = 'open';
+  private removeExecutionsTimers: number[] = [];
```
The name of this variable is an action, I'd suggest to use subject, just simple `timers` would be enough in this context.

dimadeveatii reviewed (changes requested):
I like the implementation. Please cover also with unit tests each file.

Commit:
circuit
implement factories as class (not function)
move each factory to its entity folde
change isPromise function

Commit:
circuit
add tests for CircuitState and its factory
and tests for Policies and their factory

dimadeveatii commented on lib/circuit/CircuitState/factory.ts:
```diff
@@ -0,0 +1,21 @@
+import { CircuitState } from './CircuitState';
+
+export class CircuitStateFactory implements Factory<CircuitState> {
+
```
Consecutive blank lines might fail linting.

dimadeveatii commented on lib/circuit/CircuitStateStorage/ArgumentsCircuitStateStorage.ts:
```diff
@@ -0,0 +1,22 @@
+  constructor(private readonly circuitStateFactory: CircuitStateFactory) { }
+
+  public get(args: any[]): CircuitState {
+    const key = hash(args);
```
Later, when merged with `@cache` this has to be refactored to use the `HashService` from `/utils`

dimadeveatii commented on lib/circuit/CircuitStateStorage/ArgumentsCircuitStateStorage.ts:
```diff
@@ -0,0 +1,22 @@
+
+export class ArgumentsCircuitStateStorage implements CircuitStateStorage {
+
+  private readonly argumentsStorage = new Map<string, CircuitState>();
```
Will the elements from this map ever be deleted? Aren't we introducing a memory leak here?

JohnDoePlusPlus replied:
I added logic to remove keys from this map in last commit, I hope implementation is ok

dimadeveatii commented on lib/circuit/CircuitStateStorage/factory.ts:
```diff
@@ -0,0 +1,43 @@
+
+  private instanceCircuitStateStorage(): InstanceCircuitStateStorage {
+    return new InstanceCircuitStateStorage(this.circuitStateFactory);
+  }
```
Because all of this are just one liners with suggestive class name, I'd say just to use them directly in switch and avoid functions. The functions does not tell me anything new useful.

dimadeveatii commented on lib/circuit/CircuitState/CircuitState.ts:
```diff
@@ -0,0 +1,74 @@
+      } else {
+        this.open();
+      }
+    }
```
Please move this entire block to a private method.

dimadeveatii commented on lib/circuit/CircuitState/CircuitState.ts:
```diff
@@ -0,0 +1,74 @@
+    return this;
+  }
+
+  private removeExecution(type: 'success' | 'error'): void {
```
`removeExecution` name is misleading in my opinion. It does not remove it, rather it registers it for later removal. A name like `registerCall` would be more appropriate (or maybe smth else, but not `remove`).

dimadeveatii commented on lib/circuit/Policy/factory.ts:
```diff
@@ -0,0 +1,34 @@
+
+  private ratePolicy(): RatePolicy {
+    return new RatePolicy(this.threshold);
+  }
```
Same as in other comment. Functions does not tell me anything new, so use `new` directly in switch.

dimadeveatii commented on lib/utils/isPromise.ts:
```diff
@@ -0,0 +1,5 @@
+export function isPromise(data: any): data is Promise<unknown> {
+  return data
+    && data.then && typeof data.then === 'function'
+    && data.catch && typeof data.catch === 'function';
```
`catch` is not mandatory in a promise-like construction.

Commit:
circuit
change isPromise to isPromiseLike
add more tests
remove from factories one line functions
add spies

Commit:
circuit
rename policy methods
CircuitState separate code in methods

Commit:
circuit
CircuitState separate code in methods
and rename removeExecution method to registerCall

Commit:
Merge branch 'feature/circuit' of github.com:JohnDoePlusPlus/micro-decorators into feature/circuit

Commit:
circuit
add unit test (now everything is tested)
change implementation of CircuitState and ArgumentsCircuitStateStorage
to avoid memory leaks

Commit:
circuit
improve unit tests

Commit:
circuit
improve tests

JohnDoePlusPlus committed:
circuit
improve all factories
adjust unit tests for factories
